### DISCUSSION TITLE
How to use mask in enmap ?

### DISCUSSION THREAD

**Discussion by vprint** [EnMAP-Box/enmap-box#94] [Q&A] [answered]
Hello,

I created a wetland mask from a landsat image using qgis reclassify by table tool. This reclassification allowed me to create a binary raster of wetlands area. However if i run a pca or a kmeans using the mask i got data on the masked area. 

Exemple for pca :

original data
![image](https://user-images.githubusercontent.com/42189444/193328951-3392ccf8-02ad-47b1-9884-b63203320f42.png)

wetland mask
![image](https://user-images.githubusercontent.com/42189444/193329008-053ea266-b2f6-4f9e-83f1-e768cadc619f.png)

outputed PCA (using mask)
![image](https://user-images.githubusercontent.com/42189444/193329027-fdd15122-154d-4ce0-82af-1f434212e5d5.png)

Steps :
1)	Load landsat and mask data
3)	Create unsupervised dataset and apply the wetland mask (Create Unsupervised Dataset tool)
4)	Run a PCA using the previously created dataset (Fit PCA tool)
5)	Tranform landsat raster layer using PCA .pkl

Result : 
The whole area is treated, while only unmasked area shoud be.

If i want to apply the mask, i have to use « Apply mask layer to raster layer », even if i already filled the mask in the « create unsupervised dataset » step. Is it normal ?

Here is an extract of my data if you need them
[dataset.zip](https://github.com/EnMAP-Box/enmap-box/files/9686705/dataset.zip)

**Comment by janzandr** [comment, marked as answer]
The mask that you specify here is only limitting the training samples for the PCA.
![image](https://user-images.githubusercontent.com/45030715/193367739-7b5ba432-3ea9-4291-9a93-2e0d77256d83.png)

In the Transform step, only those areas are masked, which are already masked in the raster:
![image](https://user-images.githubusercontent.com/45030715/193367982-3f1b7255-8d4d-4fc9-84a7-575499d0678e.png)

Note that instead of transforming the original raster, you could transform any other Landsat raster with the fitted PCA model. For those other rasters, the original mask would have no meaning.

Long story short. You could a) mask your original **landsat_data.tif** before drawing the samples, or b) mask the output PCA raster.

**Comment by vprint** [comment]
Thank you, If i mask my original landsat_data.tif before drawing the samples then i obtain the following message when runnning "transform raster layer tool" : 

Traceback (most recent call last):
              File "C:\Users/valen/AppData/Roaming/QGIS/QGIS3\profiles\default/python/plugins\enmapboxplugin\site-packages\typeguard\__init__.py", line 903, in wrapper
              retval = func(*args, **kwargs)
              File "C:\Users/valen/AppData/Roaming/QGIS/QGIS3\profiles\default/python/plugins\enmapboxplugin\enmapboxprocessing\algorithm\transformrasteralgorithm.py", line 100, in processAlgorithm
              Xt = dump.transformer.transform(np.transpose(X))
              File "C:\Users\valen\AppData\Roaming\Python\Python39\site-packages\sklearn\pipeline.py", line 635, in transform
              Xt = transform.transform(Xt)
              File "C:\Users\valen\AppData\Roaming\Python\Python39\site-packages\sklearn\decomposition\_base.py", line 120, in transform
              X = self._validate_data(X, dtype=[np.float64, np.float32], reset=False)
              File "C:\Users\valen\AppData\Roaming\Python\Python39\site-packages\sklearn\base.py", line 577, in _validate_data
              X = check_array(X, input_name="X", **check_params)
              File "C:\Users\valen\AppData\Roaming\Python\Python39\site-packages\sklearn\utils\validation.py", line 899, in check_array
              _assert_all_finite(
              File "C:\Users\valen\AppData\Roaming\Python\Python39\site-packages\sklearn\utils\validation.py", line 146, in _assert_all_finite
              raise ValueError(msg_err)
             ValueError: Input X contains NaN.
             PCA does not accept missing values encoded as NaN natively. For supervised learning, you might want to consider sklearn.ensemble.HistGradientBoostingClassifier and Regressor which accept missing values encoded as NaNs natively. Alternatively, it is possible to preprocess the data, for instance by using an imputer transformer in a pipeline or drop samples with missing values. See https://scikit-learn.org/stable/modules/impute.html You can find a list of all estimators that handle NaN values at the following page: https://scikit-learn.org/stable/modules/impute.html#estimators-that-handle-nan-values

Looks like a scikit issue... Workaround is to apply mask on output pca layers

**Reply by janzandr** [reply]
I can't reproduce this with my testdata. The bug is probably already fixed. I would ask you to:
1. Try it with the current dev version: 
https://bitbucket.org/hu-geomatics/enmap-box/downloads/enmapboxplugin.3.11.20221001T022425.main.zip
2. If that is not working, share the masked landsat_data.tif raster.

**Comment by vprint** [comment]
Ok it works with the dev version
